contracts/abstracted_account.algo.ts

```diff
@@ -35,35 +42,169 @@ export class AbstractedAccount extends Contract {
+      return true;
+    }
+
+    return (
+      txn.typeEnum === TransactionType.ApplicationCall &&
+      txn.applicationID === this.app &&
+      txn.numAppArgs === 1 &&
+      txn.onCompletion === 0 &&
+      txn.applicationArgs[0] === method('arc58_verifyAuthAddr()void')
+    )
+  }
+
+  private assertRekeysBack(): void {
+    let rekeysBack = false;
+    for (let i = (this.txn.groupIndex + 1); i < this.txnGroup.length; i += 1) {
+      const txn = this.txnGroup[i];
+
+      if (this.ensuresRekeyBack(txn)) {
+        rekeysBack = true;
+      }
+    }
+
+    assert(rekeysBack, 'rekey back not found');
+  }
+
   /**
-   * Ensure that by the end of the group the abstracted account has control of itself
+   * Guarantee that our txn group is valid in a single loop over all txns in the group
+   * 
+   * @param app the plugin app id being validated
+   * @param checkGlobal whether to check the global caller for method restrictions
+   * @param checkLocal whether to check the local caller for method restrictions
    */
-  private verifyRekeyToAbstractedAccount(): void {
-    let rekeyedBack = false;
+  private assertValidGroup(app: AppID, methodOffsets: uint64[], checkGlobal: boolean, checkLocal: boolean): CallerUsed {
+    const gkey: PluginsKey = { application: app, allowedCaller: Address.zeroAddress };
+    const key: PluginsKey = { application: app, allowedCaller: this.txn.sender };
+
+    let rekeysBack = false;
+
+    const globalRestrictions = checkGlobal && this.plugins(gkey).size > 29;
+    const localRestrictions = checkLocal && this.plugins(key).size > 29;
+    let methodIndex = 0;
+    let callerUsed: CallerUsed = {
+      global: checkGlobal && !globalRestrictions,
+      local: checkLocal && !localRestrictions,
+    };
 
-    for (let i = this.txn.groupIndex + 1; i < this.txnGroup.length; i += 1) {
+    for (let i = (this.txn.groupIndex + 1); i < this.txnGroup.length; i += 1) {
       const txn = this.txnGroup[i];
 
-      // The transaction is an explicit rekey back
-      if (txn.sender === this.app.address && txn.rekeyTo === this.app.address) {
-        rekeyedBack = true;
-        break;
+      if (this.ensuresRekeyBack(txn)) {
+        rekeysBack = true;
       }
 
-      // The transaction is an application call to this app's arc58_verifyAuthAddr method
-      if (
-        txn.typeEnum === TransactionType.ApplicationCall &&
-        txn.applicationID === this.app &&
-        txn.onCompletion === 0 && // OnCompletion.NoOp
-        txn.numAppArgs === 1 &&
-        txn.applicationArgs[0] === method('arc58_verifyAuthAddr()void')
-      ) {
-        rekeyedBack = true;
-        break;
+      // we dont need to check method restrictions at all if none exist
+      // & skip transactions that aren't relevant
+      if ((!globalRestrictions && !localRestrictions) || this.shouldSkipMethodCheck(txn, app)) {
+        continue;
       }
+
+      const globalValid = (
```

Comment:
This is pretty tough to read. I think it's preferable to put multiple bools in a var first that describes each condition.